messaging/app/src/main/java/com/google/firebase/quickstart/fcm/kotlin/MainActivity.kt

```diff
@@ -80,25 +88,46 @@ class MainActivity : AppCompatActivity() {
         binding.logTokenButton.setOnClickListener {
             // Get token
             // [START log_reg_token]
-            Firebase.messaging.getToken().addOnCompleteListener(OnCompleteListener { task ->
-                if (!task.isSuccessful) {
-                    Log.w(TAG, "Fetching FCM registration token failed", task.exception)
-                    return@OnCompleteListener
-                }
-
-                // Get new FCM registration token
-                val token = task.result
-
+            GlobalScope.launch {
+                val token = getAndStoreRegToken()
                 // Log and toast
-                val msg = getString(R.string.msg_token_fmt, token)
-                Log.d(TAG, msg)
-                Toast.makeText(baseContext, msg, Toast.LENGTH_SHORT).show()
-            })
+                Log.d(TAG, "on click listener: $token")
+                val deviceToken = hashMapOf(
+                    "token" to token,
+                    "timestamp" to FieldValue.serverTimestamp(),
+                )
+
+                // Get user ID from Firebase Auth or your own server
+                Firebase.firestore.collection("fcmTokens").document("myuserid")
+                    .set(deviceToken)
+            }
             // [END log_reg_token]
         }
 
         Toast.makeText(this, "See README for setup instructions", Toast.LENGTH_SHORT).show()
         askNotificationPermission()
+
+        // In the app’s first Activity
+        val preferences = this.getPreferences(Context.MODE_PRIVATE)
+        val lastRefreshLong = preferences.getLong("lastRefreshDate", -1)
+        Log.d(TAG, "last refresh in preferences: $lastRefreshLong")
+        GlobalScope.launch {
+            val document = Firebase.firestore.collection("refresh").document("refreshDate").get().await()
+            val updatedTime = (document.data!!["lastRefreshDate"] as Timestamp).seconds * 1000
+            Log.d(TAG, "updatedTime: $updatedTime from Firestore")
+            val lastRefreshDate = Date(if (lastRefreshLong == -1L) updatedTime else lastRefreshLong)
+            val c = Calendar.getInstance()
+            c.time = lastRefreshDate
+            c.add(Calendar.DATE, 30)
```

Comment:
nit: In Kotlin you can use the `apply()` function to avoid repeating the `c` variable in every line:
```suggestion
            val c = Calendar.getInstance().apply {
                time = lastRefreshDate
                add(Calendar.DATE, 30)
            }
```